From 98e3ac26bbdf8e1b3c6ada1160a566bf2e12963a Mon Sep 17 00:00:00 2001
From: Devarsh Thakkar <devarsh.thakkar@xilinx.com>
Date: Wed, 7 Aug 2019 11:44:35 -0700
Subject: [PATCH 0656/1852] v4l: xilinx: dma: Perform early callback on queued
 buffers

commit 2bfdadf589853910130a046c5f9d3ed2996c90ab from
https://github.com/Xilinx/linux-xlnx.git

This patch enables the early release of queued buffers to application
in case of low latency capture.

Signed-off-by: Devarsh Thakkar <devarsh.thakkar@xilinx.com>
Signed-off-by: Satish Kumar Nagireddy <satish.nagireddy.nagireddy@xilinx.com>
Reviewed-by: Hyun Kwon <hyun.kwon@xilinx.com>
State: pending
Signed-off-by: Yaliang Wang <Yaliang.Wang@windriver.com>
---
 drivers/media/platform/xilinx/xilinx-dma.c | 6 ++++++
 1 file changed, 6 insertions(+)

diff --git a/drivers/media/platform/xilinx/xilinx-dma.c b/drivers/media/platform/xilinx/xilinx-dma.c
index 8299cda085d6..1a8533a927ba 100644
--- a/drivers/media/platform/xilinx/xilinx-dma.c
+++ b/drivers/media/platform/xilinx/xilinx-dma.c
@@ -631,6 +631,12 @@ static void xvip_dma_buffer_queue(struct vb2_buffer *vb)
 	list_add_tail(&buf->queue, &dma->queued_bufs);
 	spin_unlock_irq(&dma->queued_lock);
 
+	/* Low latency capture: Early release of buffers to application */
+	if (dma->low_latency_cap) {
+		desc->callback = NULL;
+		xvip_dma_complete(buf);
+	}
+
 	dmaengine_submit(desc);
 
 	if (vb2_is_streaming(&dma->queue))
-- 
2.31.1

